async/cohttp_async.ml

```diff
@@ -156,25 +156,60 @@ end
 
 module Client = struct
 
+  let read_request ic =
+    Response.read ic >>| function
+    | `Eof -> failwith "Connection closed by remote host"
+    | `Invalid reason -> failwith reason
```

Comment:
Actually, I don't think that my previous characterization of the issue is correct. The current implementation makes it impossible to read all responses that were successfully written to the pipe.

Some care will be necessary however because an exception might be thrown before the corresponding response is actually processed in the pipe. But you should be able to catch it and iterate over the pipe regardless.